src/goto-instrument/contracts/utils.h

```diff
@@ -110,6 +110,23 @@ void insert_before_swap_and_advance(
   goto_programt::targett &target,
   goto_programt &payload);
 
+/// \brief Insert a goto instruction before a target instruction iterator
+///        and update targets of all jumps that points to the iterator to
+///        jumping to the inserted instruction.
+///
+/// This function inserts a instruction `i` into a destination program
+/// `destination` immediately before a specified instruction iterator `target`.
+/// After insertion, update all jumps that pointing to `target` to jumping to
+/// `i` instead.
+///
+/// \param destination: The destination program for inserting the `i`.
+/// \param target: The instruction iterator at which to insert the `i`.
+/// \param i: The goto instruction to be inserted into the `destination`.
+void insert_before_and_update_jumps(
+  goto_programt &destination,
+  goto_programt::targett &target,
+  const goto_programt::instructiont &i);
```

Comment:
Isn't that what `insert_before_swap` (in its several variants, from `goto-programs/goto_program.h`) does?!

Reply:
Please (mostly) disregard: it seems we need both (at least for now). So maybe the comment could clarify how this relates to `insert_before_swap`.

Reply:
I added comments to explain that `insert_before_swap` will invalidate `target`.
If we use `insert_before_swap_and_advance` in this PR, we have to also update `loop_content` (https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/contracts/contracts.cpp#L870): the original `loop_head` will point to the new skip instruction after the insertion so that we have to erase it from `loop_content` and insert the new `loop_head` (return by `insert_before_swap_and_advance`) into `loop_content`.